### PR TITLE
Fix showSnackBar can't access useMaterial3 from the theme

### LINKED ISSUE #115924: Can't access `useMaterial3` from the theme in the `showSnackBar` method to conditionally change `SnackBar` animation for Material 3.
While working on https://github.com/flutter/flutter/pull/115750, I was attempting to conditionally change the duration of the SnackBar animation, depending on whether the theme was Material 3 or not. 

The Scaffold creates the animation controller for its snackbars in the `ScaffoldMessengerState` method `showSnackBar()`:
```
 _snackBarController ??= SnackBar.createAnimationController(vsync: this)
```
[(scaffold.dart)](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/scaffold.dart#L295)

`createAnimationController` is a static method of `SnackBar`:
```
  static AnimationController createAnimationController({ required TickerProvider vsync }) {
    return AnimationController(
      duration: _snackBarTransitionDuration,
      debugLabel: 'SnackBar',
      vsync: vsync,
    );
  }
 ``` 
[(snack_bar.dart)](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/snack_bar.dart#L363)

Ideally we could pass an optional `useMaterial3` parameter to this method, and change the duration accordingly. 

However, when running a Material 3 example (like `snackbar.2.dart` in the above mentioned PR) I can't seem to access a value of `Theme.of(context).useMaterial3 = true` from `ScaffoldMessengerState`, though I can from `ScaffoldState`.

Perhaps there's a trivial solution to this, but I need some insight from someone with more Scaffold familiarity.

### PR DESCRIPTION
**Pull request by bleroux**
## Description

This PR makes it possible for the `MaterialApp` built in `ScaffoldMessenger` state to access the ambient theme.

Before this PR, the built in messenger was above the theme.
After this PR, the build in messenger is below the theme.

## Related Issue

Fixes [Can't access useMaterial3 from the theme in the showSnackBar method](https://github.com/flutter/flutter/issues/115924)

## Tests

Adds 1 test.

### CODE REVIEW

File: packages/flutter/test/material/scaffold_test.dart [hunk 1]
```diff
@@ -2867,7 +2867,47 @@ void main() {
...
+            );
+          },
+        ),
+      ),
+    ));
+
+    await tester.tap(find.byKey(tapTarget));
+    await tester.pump();
+
+    final ThemeData messengerTheme = Theme.of(scaffoldMessenger!.context);
```
Comment (nate-thegrate) [inline packages/flutter/test/material/scaffold_test.dart:2905]: I believe we could make use of MaterialApp's [scaffoldMessengerKey](https://main-api.flutter.dev/flutter/material/MaterialApp/scaffoldMessengerKey.html) parameter here:

```suggestion
    await tester.pumpWidget(
      MaterialApp(
        scaffoldMessengerKey: scaffoldMessengerKey,
        theme: customTheme,
        home: const SizedBox.shrink(),
      ),
    );

    final ThemeData messengerTheme = Theme.of(scaffoldMessengerKey.currentContext!);
```

File: packages/flutter/test/material/scaffold_test.dart [hunk 2]
```diff
@@ -2867,7 +2867,47 @@ void main() {
       expect(tester.takeException(), isNull);
   });
 
-   testWidgets('ScaffoldMessenger showSnackBar default animation', (WidgetTester tester) async {
+  // Regression test for https://github.com/flutter/flutter/issues/115924.
+  testWidgets('Default ScaffoldMessenger can access ambient theme', (WidgetTester tester) async {
+    ScaffoldMessengerState? scaffoldMessenger;
```
Comment (nate-thegrate) [inline packages/flutter/test/material/scaffold_test.dart:2872]:
```suggestion
    final GlobalKey<ScaffoldMessengerState> scaffoldMessengerKey = GlobalKey<ScaffoldMessengerState>();
```

File: packages/flutter/test/material/scaffold_test.dart [hunk 3]
```diff
@@ -2867,7 +2867,47 @@ void main() {
       expect(tester.takeException(), isNull);
   });
 
-   testWidgets('ScaffoldMessenger showSnackBar default animation', (WidgetTester tester) async {
```
Comment (nate-thegrate) [inline packages/flutter/test/material/scaffold_test.dart:2870]: I had some trouble trying to see what was changed about this line… thanks for fixing the indent!

**Review by nate-thegrate — COMMENTED** [review]
It's super great to see the simple fix, thank you!

I had an idea for the test and would love to hear your thoughts :)

**Comment by nate-thegrate** [comment]
> **bleroux** requested a review from **Piinks**

(Kate is out this week btw)

File: packages/flutter/test/material/scaffold_test.dart (hunk 1 above)
Comment (bleroux) [reply packages/flutter/test/material/scaffold_test.dart:2905]: Nice! I will make the change tomorrow morning (it is the evening here).

File: packages/flutter/lib/src/material/app.dart [hunk 4]
```diff
@@ -986,6 +986,16 @@ class _MaterialAppState extends State<MaterialApp> {
...
       );
     }
+
+    childWidget = ScaffoldMessenger(
+      key: widget.scaffoldMessengerKey,
+      child: DefaultSelectionStyle(
+        selectionColor: effectiveSelectionColor,
+        cursorColor: effectiveCursorColor,
+        child: childWidget,
+      ),
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/app.dart:996]: Probably not in the scope of this PR, but I noticed that the `Theme` widget already builds a `DefaultSelectionStyle`:

https://github.com/flutter/flutter/blob/16622e67a051a46dc0636df0e06322644411b92d/packages/flutter/lib/src/material/theme.dart#L132-L142

Would be nice to remove the duplicate widget here at some point :)

**Review by nate-thegrate — APPROVED** [review]
Flutter_LGTM

Thank you very much!

**Pull request merged**